File: .portkit/scripts/typescript/fetch-upstream.ts
// Fetch Upstream - Ingestion Strategy
// Purpose: Manages .portkit-cache shadow repo (clone/fetch).

import * as fs from 'fs';
import * as path from 'path';
import { execSync } from 'child_process';

const CONFIG_PATH = path.join('.portkit', 'config.json');
const CACHE_DIR = '.portkit-cache';

interface Config {
    upstream_url: string;
}

function loadConfig(): Config {
    if (!fs.existsSync(CONFIG_PATH)) {
        console.error(`Error: Config file not found at ${CONFIG_PATH}`);
        process.exit(1);
    }
    return JSON.parse(fs.readFileSync(CONFIG_PATH, 'utf-8'));
}

async function main() {
    const args = process.argv.slice(2);
    const targetRef = args[0]; // e.g., 'main', 'v1.0.0', 'sha123'

    if (!targetRef) {
        console.error("Usage: npx ts-node fetch-upstream.ts <git-ref>");
        process.exit(1);
    }

    const config = loadConfig();
    const repoUrl = config.upstream_url;

    if (!fs.existsSync(CACHE_DIR)) {
        console.log(`Cache missing. Cloning upstream from ${repoUrl}...`);
        try {
            execSync(`git clone ${repoUrl} ${CACHE_DIR}`, { stdio: 'inherit' });
        } catch (e) {
            console.error("Failed to clone upstream repository.");
            process.exit(1);
        }
    } else {
        console.log(`Cache exists. Fetching updates...`);
        try {
            execSync(`git fetch --all`, { cwd: CACHE_DIR, stdio: 'inherit' });
        } catch (e) {
            console.error("Failed to fetch updates.");
            process.exit(1);
        }
    }

    console.log(`Checking out ${targetRef}...`);
    try {
        execSync(`git checkout ${targetRef}`, { cwd: CACHE_DIR, stdio: 'inherit' });
        // Optional: Pull if it's a branch, but checkout is safer for detached refs
        try {
            execSync(`git pull origin ${targetRef}`, { cwd: CACHE_DIR, stdio: 'ignore' });
        } catch (e) {
            // Ignore pull errors (e.g. detached head, tag)
        }

        console.log(`Successfully checked out ${targetRef} in ${CACHE_DIR}`);
    } catch (e) {
        console.error(`Failed to checkout ${targetRef}`);
        process.exit(1);
    }
}

main().catch(err => {
    console.error(err);
    process.exit(1);
});

File: .portkit/scripts/python/fetch-upstream.py
import os
import subprocess
import argparse
import json
import sys
from pathlib import Path

# Configuration
PORTKIT_ROOT = Path(__file__).resolve().parent.parent.parent
CONFIG_FILE = PORTKIT_ROOT / "config.json"
CACHE_DIR = PORTKIT_ROOT.parent / ".portkit-cache"

def load_config():
    if not CONFIG_FILE.exists():
        print(f"Error: Config file not found at {CONFIG_FILE}")
        sys.exit(1)
    with open(CONFIG_FILE, "r") as f:
        return json.load(f)

def run_git(args, cwd):
    try:
        subprocess.run(["git"] + args, cwd=cwd, check=True, capture_output=True, text=True)
    except subprocess.CalledProcessError as e:
        print(f"Git Error: {e.stderr}")
        sys.exit(1)

def main():
    parser = argparse.ArgumentParser(description="Fetch and cache upstream repository.")
    parser.add_argument("--ref", help="Target git reference (tag/branch/commit)", default="main")
    args = parser.parse_args()

    config = load_config()
    upstream_url = config.get("upstream_url")
    
    if not upstream_url:
        print("Error: 'upstream_url' not defined in config.json")
        sys.exit(1)

    # Ensure cache directory exists
    if not CACHE_DIR.exists():
        print(f"Cloning upstream from {upstream_url}...")
        # Clone straight into .portkit-cache
        # We need to create the parent dir context if .portkit-cache doesn't exist?
        # git clone creates the dir.
        try:
            subprocess.run(["git", "clone", upstream_url, str(CACHE_DIR)], check=True, capture_output=False)
        except subprocess.CalledProcessError as e:
            print(f"Clone Failed: {e}")
            sys.exit(1)
    else:
        print("Fetching latest from upstream...")
        run_git(["fetch", "--all"], cwd=CACHE_DIR)

    # Checkout Target
    print(f"Checking out {args.ref}...")
    run_git(["checkout", args.ref], cwd=CACHE_DIR)
    
    # Enable pulling if it's a branch
    try:
        run_git(["pull"], cwd=CACHE_DIR)
    except:
        pass # Detached head or tag

    print(f"Success. Upstream available at: {CACHE_DIR}")

if __name__ == "__main__":
    main()
